opentracing-shim/src/opentracingshim/__init__.py

```diff
@@ -0,0 +1,241 @@
+
+from contextlib import contextmanager
+
+import opentracing
+
+from opentelemetry.trace import Event
+from opentracingshim import util
+
+
+def create_tracer(tracer, scope_manager=None):
+    return TracerWrapper(tracer, scope_manager)
+
+
+class SpanContextWrapper(opentracing.SpanContext):
+    def __init__(self, otel_context):
+        self._otel_context = otel_context
+
+    def unwrap(self):
+        """Returns the wrapped OpenTelemetry `SpanContext` object."""
+
+        return self._otel_context
+
+    @property
+    def baggage(self):
+        raise NotImplementedError
+        # TODO: Implement.
+
+
+class SpanWrapper(opentracing.Span):
+    def __init__(self, tracer, context, span):
+        self._otel_span = span
+        super().__init__(tracer, context)
+
+    def unwrap(self):
+        """Returns the wrapped OpenTelemetry `Span` object."""
+
+        return self._otel_span
+
+    def set_operation_name(self, operation_name):
+        self._otel_span.update_name(operation_name)
+
+        # Return self for call chaining.
+        return self
+
+    def finish(self, finish_time=None):
+        end_time = finish_time
+        if end_time is not None:
+            end_time = util.time_seconds_to_ns(finish_time)
+        self._otel_span.end(end_time=end_time)
+
+    def set_tag(self, key, value):
+        self._otel_span.set_attribute(key, value)
+
+        # Return self for call chaining.
+        return self
+
+    def log_kv(self, key_values, timestamp=None):
+        if timestamp is not None:
+            event_timestamp = util.time_seconds_to_ns(timestamp)
+        else:
+            event_timestamp = util.time_ns()
+
+        event_name = util.event_name_from_kv(key_values)
+        event = Event(event_name, event_timestamp, key_values)
+        self._otel_span.add_lazy_event(event)
+
+        # Return self for call chaining.
+        return self
+
+    def set_baggage_item(self, key, value):
+        raise NotImplementedError
+        # TODO: Implement.
+
+    def get_baggage_item(self, key):
+        raise NotImplementedError
+        # TODO: Implement.
+
+    # TODO: Verify calls to deprecated methods `log_event()` and `log()` on
+    # base class work properly (it's probably fine because both methods call
+    # `log_kv()`).
+
+
+class ScopeWrapper(opentracing.Scope):
+    def close(self):
+        self._span.finish()
+        # TODO: Set active span on OpenTelemetry tracer.
+        # https://github.com/open-telemetry/opentelemetry-python/issues/161#issuecomment-534136274
+
+
+class ScopeManagerWrapper(opentracing.ScopeManager):
+    def __init__(self, tracer):
+        # The only thing the `__init__()` method on the base class does is
+        # initialize `self._noop_span` and `self._noop_scope` with no-op
+        # objects. Therefore, it doesn't seem useful to call it.
+        # pylint: disable=super-init-not-called
+        self._tracer = tracer
+
+    @contextmanager
+    def activate(self, span, finish_on_close):
```

Comment:
I don't think this method does what you want. After `@contextmanager` does it's magic, we have a method that returns a ContextManager (without any additional methods) that returns a `ScopeWrapper` (which is also a ContextManager) from it's `__enter__`. This means that doing `with myScopeManagerWrapper.activate(foo):` will only call `__exit__` on the outer context manager, thus calling `__exit__` on the context manager returned by `use_span`. If you use `with with myScopeManagerWrapper.activate(foo) as outer, outer:` you call `__exit__` on both.

I guess we should resolve this by:
* Not using `@contextmanager` here
* Directly returning (instead of yielding) the `ScopeWrapper`
* The scope wrapper calls or receives the context manager returned by `otel_tracer.use_span` in it's `__init__`.
* The scope wrapper manually calls `__enter__` and `__exit__` on that context manager from its corresponding methods.

Reply:
Great feedback @Oberon00, thanks. I think you are right.
Fixed now (hopefully).